### ISSUE TITLE
Ranging modes options

### ISSUE THREAD

**Issue by psyhtest** [mlcommons/power-dev#270]
At the moment, the ranging run takes the same time (at least 10 minutes) as the testing run, which is wasteful.

I would like to propose the following options:
- None or Manual, when the ranging run is not performed, while the range is set explicitly. This useful when the range is known a priori. For example, [the official Raspberry Pi 4 power supply](https://www.raspberrypi.com/products/type-c-power-supply/) is rated at 15 W.
- Reduced, when the ranging run is allowed to take less time than the testing run, e.g. about one minute.

**Comment by arjunsuresh** [comment]
Just to understand the implication of non-precise manual setting,

- if I manually put 1000W as the limit and the actual power usage during testing run is say 1200W or 800W, will there be any error in the power measurement as compared to precisely setting the limit?

**Comment by psyhtest** [comment]
There will be an error if the actual power usage is 1200W. There will be _no_ error if the actual power usage is 800W.

**Comment by arjunsuresh** [comment]
Thanks @psyhtest. And the error when power usage is 1200W will come in logs marking the run as invalid? In other words can we be sure that no run pass as a valid run with an imprecise power measurement due to a wrong manual input?

**Comment by psyhtest** [comment]
I'm sure it is the case. When the range is set incorrectly, there will be errors like:
> '02-21-2022 19:42:02.524: ERROR: Bad watts reading nan from WT310' in ptd_log.txt during ranging stage. Treated as WARNING

**Comment by arjunsuresh** [comment]
Thank you @psyhtest. In that case I see absolutely no problem in using "None" or "Manual" mode. It can be an option which experienced users can use at the risk of handling the errors themselves and the measured values are having the same precision as with a full ranging mode run. 

For a 1-2 minutes ranging run I can think of some issues

- Typically the accelerator will be at full power usage by 1 minute
- But the system fans may not have reached full speed (supposing the fan speed is dynamic). This can cause a lot of increase in power usage
- Suppose X is the peak power usage for a 1 minute ranging run.  Can we use X + threshold as the limit for the testing run where the threshold is a value good enough to ensure no failure for all the power ranges?

I would clearly favor a "None" or "Manual" mode for experienced users as it is simpler (at least for repeated users) and  there is no precision loss.

**Comment by psyhtest** [comment]
I agree @arjunsuresh. The set of analyzer ranges is quite discreet e.g. for WT310/WT330: 0.5A/1A/2A/5A/10A/20A. If one range fails, the submitter should be able to switch to the next one with little trouble.

**Comment by arjunsuresh** [comment]
That's great. So any submitter can do a ranging run - determine the max power usage and then use that for all further runs. This effectively saves 50% of the runtime for all power submissions and if we are getting that with 0 loss of precision, we should go for it.

**Comment by araghun** [comment]
This PR was bought up in the 10/25 Power WG meeting.
tl;dr is @psyhtest were create 2 PRs from this - one for Open Division (or Low Power SUTs) and other for Closed Divisions. The Closed division PR will need data to reach a consensus.  The open division PR is relatively straight forward to approve and close.

Some of the questions/discussion points from the Power WG meetings minutes are:

- Question: Can there be a scenario where there are no errors reported, yet the data is inaccurate? Any results to show that? This would require us to set up Manual mode and then test it.
- Thought: Another alternative is to reduce the ranging step to a lower time bound (say currently it takes 2x time for Ranging and Testing. Reduce this to 1.2x pr 1.3x)
- Question: Need data to back the proposal. What is the tradeoff that is being done (Time taken vs possible error bar)
- Question: Do we need to do this manual setting for all scenarios and all workloads? In that way, are we really saving time?
- Proposal: Instead of running Ranging and Testing as 2 separate steps, can we just combine into 1 performance step.

**Comment by arjunsuresh** [comment]
Just adding the updates from the discussion with the SPEC PTDaemon team:

1. Relying on the data from PTDaemon logs is not recommended
2. There is an "Uncertainty" command which can be reliably used to measure the uncertainty of measurement

We are still awaiting a confirmation from SPEC team regarding if "Uncertainty command" can be used to optionally avoid the Ranging mode run. This should answer Question 1 in the above comment by @araghun and if the response is favorable we do not need to see points 2 and 3.

 "Do we need to do this manual setting for all scenarios and all workloads? In that way, are we really saving time?"

Yes, we need it for all scenarios and all workloads.  And in fact we reduce the runtimes for all of them by at least 50%. Why "at least 50%" is because "Server" scenario can cause performance failure on either the "ranging" mode or the "testing mode" and either case a full rerun is needed. 

"Proposal: Instead of running Ranging and Testing as 2 separate steps, can we just combine into 1 performance step."

If this is technically possible this will be very useful. The first 5 minutes of the performance run can done in ranging mode and the remaining time (guaranteed to be at least 5 minutes) can be used for testing. If this can be done we can even ignore all other questions here. @psyhtest do you agree?

**Comment by arjunsuresh** [comment]
Reply from SPEC team:
"I don't think it really matters whether the ranges are set algorithmically by software based on a test run or by a user with knowledge of a system's power draw editing a configuration file. The important thing is the supported commands are used to retrieve results. If only option 1 uses the Uncertainty command, then it is more likely to produce valid results."

This clearly shows that doing a manual setting for testing mode is 100% accurate if we use the Uncertainty command as documented. I can do this change is anyone is interested and the Power WG is accepting this. Unfortunately I do not have time to try any other methodology.

**Comment by arjunsuresh** [comment]
I have added a [PR](https://github.com/octoml/power-dev/commit/bec5a4d26a9ffa7615c0745e043f442088468e7d) effecting the needed change for manul setting of volt/ampere levels. Some sample test outputs are also added [here](https://github.com/mlcommons/power-dev/pull/273#issuecomment-1364717406). Uncertainty command is capturing wrong inputs.
